### PR TITLE
Add Write + Conflict Resolution

### PR DESCRIPTION
**Pull request by matt-ramotar**
Please see our contributing guidelines (contributing.md) primarily make sure to sign our cla as we cannot accept code externally without a signed cla

https://opensource.dropbox.com/cla/

### CODE REVIEW

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/ItemValidator.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.mobilenativefoundation.store.store5.impl.RealItemValidator
+
+/**
+ * Enables custom validation of [Store] items.
+ * @see [ReadRequest]
+ */
+interface ItemValidator<CommonRepresentation : Any> {
+    /**
+     * Determines whether a [Store] item is valid.
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/ItemValidator.kt:11]: change comment to a Store's local item (Source of Truth or Memory Cache)

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/ItemValidator.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+     * If invalid, [Market] will get the latest network value using [NetworkFetcher].
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/ItemValidator.kt:12]: WriteStore (search replace Market everywhere)

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/MutableStore.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package org.mobilenativefoundation.store.store5
+
+import kotlinx.coroutines.flow.Flow
+
+interface MutableStore<Key : Any, CommonRepresentation : Any, NetworkWriteResponse : Any> : Store<Key, CommonRepresentation> {
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/MutableStore.kt:5]: I think we should consider not encoding the network write response type into the store's signature

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/MutableStore.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    @ExperimentalStoreApi
+    fun stream(stream: Flow<StoreWriteRequest<Key, CommonRepresentation, NetworkWriteResponse>>): Flow<StoreWriteResponse<NetworkWriteResponse>>
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/MutableStore.kt:7]: rename param to requestStream

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreBuilder.kt [hunk 5]
```diff
@@ -54,22 +53,61 @@ interface StoreBuilder<Key : Any, Output : Any> {
          *
          * @param fetcher a [Fetcher] flow of network records.
          */
-        @OptIn(ExperimentalTime::class)
-        fun <Key : Any, Output : Any> from(
-            fetcher: Fetcher<Key, Output>
-        ): StoreBuilder<Key, Output> = RealStoreBuilder(fetcher)
+        fun <Key : Any, CommonRepresentation : Any> from(
+            fetcher: Fetcher<Key, *>,
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreBuilder.kt:57]: I think you're losing type safety here

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreBuilder.kt [hunk 6]
```diff
@@ -54,22 +53,61 @@ interface StoreBuilder<Key : Any, Output : Any> {
...
+    override fun cachePolicy(memoryPolicy: MemoryPolicy<Key, CommonRepresentation>?): MutableStoreBuilder<Key, NetworkRepresentation, CommonRepresentation, SourceOfTruthRepresentation, NetworkWriteResponse>
+
+    /**
+     * @see [StoreBuilder.disableCache]
+     */
+    override fun disableCache(): MutableStoreBuilder<Key, NetworkRepresentation, CommonRepresentation, SourceOfTruthRepresentation, NetworkWriteResponse>
+
+    fun converter(converter: StoreConverter<NetworkRepresentation, CommonRepresentation, SourceOfTruthRepresentation>): MutableStoreBuilder<Key, NetworkRepresentation, CommonRepresentation, SourceOfTruthRepresentation, NetworkWriteResponse>
+
+    companion object {
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreBuilder.kt:94]: I think we should have a factory that looks something like 

`store.AsMutableStore(updater, bookkeeper)`

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreWriteRequest.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package org.mobilenativefoundation.store.store5
+
+import kotlinx.datetime.Clock
+import org.mobilenativefoundation.store.store5.impl.OnStoreWriteCompletion
+import org.mobilenativefoundation.store.store5.impl.RealStoreWriteRequest
+
+interface StoreWriteRequest<Key : Any, CommonRepresentation : Any, NetworkWriteResponse : Any> {
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/StoreWriteRequest.kt:7]: I think eventually we should collapse to

Store.Request.Write and Store.Request.Read 

same with the response

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt [hunk 8]
```diff
@@ -87,18 +92,22 @@ internal class FetcherController<Key : Any, Input : Any, Output : Any>(
                  */
                 piggybackingDownstream = true,
                 onEach = { response ->
-                    response.dataOrNull()?.let { input ->
-                        sourceOfTruth?.write(key, input)
+                    response.dataOrNull()?.let { networkRepresentation ->
+                        val input =
+                            networkRepresentation as? CommonRepresentation ?: converter?.fromNetworkRepresentationToCommonRepresentation(networkRepresentation)
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt:97]: maybe move the noop logic to the converter

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Read.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package org.mobilenativefoundation.store.store5
+
+import kotlinx.coroutines.flow.Flow
+
+interface Read {
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Read.kt:5]: @digitalbuddha - I separated into use cases so MutableStore could (1) have delegate while (2) having a generic `stream` method. Thoughts?

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/RealMutableStore.kt [hunk 10]
```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright 2019 Google LLC
```
Comment (digitalbuddha) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/RealMutableStore.kt:2]: you dont work at google do you? :-P

Comment (matt-ramotar) [reply store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/RealMutableStore.kt:2]: Haha will fix

**Review by digitalbuddha — DISMISSED** [review]

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Fetcher.kt [hunk 11]
```diff
@@ -25,6 +25,8 @@ interface Fetcher<Key : Any, NetworkRepresentation : Any> {
      */
     operator fun invoke(key: Key): Flow<FetcherResult<NetworkRepresentation>>
 
+    interface Converter
```
[the remark this answers is not in the document]
Comment (matt-ramotar) [reply store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Fetcher.kt]: Removed

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt (hunk 8 above)
Comment (matt-ramotar) [reply store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt:97]: Will address separately tomo #498

**Review by digitalbuddha — APPROVED** [review]

**Pull request merged**